---
title: "6Sept24_DE_postcop"
output: html_document
date: "2024-09-07"
---

```{r}
library(tidyverse)
library(lattice)
library(clusterProfiler)
library(enrichplot)
library(edgeR)
library(ashr)
library(pheatmap)
library(RColorBrewer)
```

#load count data 
```{r}
load("anubis_featurecounts.Rdata")
counts <- as.data.frame(anubisfc$counts)
colnames(counts)<-gsub(".anubis.useqs.merged.bam", "", colnames(counts)) 
```

#Find best postcop samples
```{r}
CountsperSampleType<-read.csv("DE_sample_metadata.csv")

Postcop<-subset(CountsperSampleType, Postcop == "Y")
Postcop_IDs<- unique(Postcop$FemaleID)

Postcop_5M <- Postcop %>%
  dplyr::filter(uniquely_mapped_total > 5000000) %>%
  filter(phasecode == "F" | phasecode == "PreF" | phasecode == "PostF") %>%
  group_by(FemaleID)

Noncop_5M <- CountsperSampleType %>%
  filter(Postcop == "N") %>%
  filter(phasecode == "F" | phasecode == "PreF" | phasecode == "PostF") %>%
  filter(uniquely_mapped_total > 5000000) %>%
  filter(FemaleID %in% Postcop_IDs) 

Postcop_DE_samples<-rbind(Postcop_5M, Noncop_5M)
Postcop_DE_samplesPostcop <- as.vector(Postcop_DE_samples$Postcop)
```

#filter counts for best samples
```{r}
countspostcop_de <- counts[,Postcop_DE_samples$Sample_ID]
```


#Create DGEList (Postcop + PreF and F noncop)
```{r}
DGEList= NULL
DGEList <- DGEList(counts=countspostcop_de, genes=anubisfc$annotation[,c("GeneID","Length")], group=factor(Postcop_DE_samplesPostcop))
```

#Remove low expressed genes (>10cpm in at least half of samples)
```{r}
keep <- rowSums(cpm(DGEList) > 10) >= 28
FilteredDGEList <- DGEList[keep,]
```

#Remove unplaced scaffolds (genes starting with "LOC")
```{r}
FilteredDGEList$counts<- FilteredDGEList$counts[!grepl("LOC", rownames(FilteredDGEList$counts)), ]
FilteredDGEList$genes<- FilteredDGEList$genes[!grepl("LOC", rownames(FilteredDGEList$genes)), ]
```

#Remove genes that are highly expressed in semen
```{r}
semengenevector<- read.table("semengenevector_19Jan25.txt")
FilteredDGEList <- FilteredDGEList[!rownames(FilteredDGEList) %in% semengenevector, ]
```

#Reset the library sizes
```{r}
FilteredDGEList$samples$lib.size <- colSums(FilteredDGEList$counts)
```


################
# Differential expression analyses
################


#Create vectors with metadata information
```{r}
Postcop <- as.vector(Postcop_DE_samples$Postcop)
Phase <- as.vector(Postcop_DE_samples$phasecode)
Postcop_DE_samples$male_presence<-gsub(" ", "", Postcop_DE_samples$male_presence)
MaleStatus <- as.vector(Postcop_DE_samples$male_presence)
RIN <- as.vector(Postcop_DE_samples$RIN)
DyadID<-as.factor(as.vector(Postcop_DE_samples$DyadID))
```

```{r}
DGE_counts_norm <- calcNormFactors(FilteredDGEList)
design <- model.matrix(~ Postcop + RIN + MaleStatus + Phase)
voom_matrix <- voomWithQualityWeights(counts = DGE_counts_norm, design) # normalizes counts
dupcor <- duplicateCorrelation(voom_matrix, design, block=DyadID)
fit <- lmFit(voom_matrix, design, block=DyadID, correlation=dupcor$consensus)# runs the linear model
fit2 <- eBayes(fit, robust=TRUE)
```

```{r}
write.table(voom_matrix$E, "postcop_normcounts.txt", sep="\t", quote=F)
```


#model results
```{r}
beta <- fit2$coefficients[, "PostcopY"]
se <- fit2$stdev.unscaled[, "PostcopY"] * fit2$sigma
pvalues<- as.data.frame(fit2$p.value)[, "PostcopY"]

#run ashr to shrink estimates
ash_fit <- ash(beta, se)

# Extract shrunken estimates
pm <- get_pm(ash_fit)
psd <- as.data.frame(get_psd(ash_fit))
lfsr <- as.data.frame(get_lfsr(ash_fit))

postcopy<-as.data.frame(cbind(pm, psd, lfsr))
colnames(postcopy)<- c("pm", "SE", "lfsr")
rownames(postcopy)<-rownames(fit2)

#subset to significant genes
sig<-subset(postcopy, lfsr < 0.1); dim(sig) #
sig5<-subset(lfsr, get_lfsr(ash_fit) < 0.05);nrow(sig5) #715
sig01<-subset(lfsr, get_lfsr(ash_fit) < 0.01); nrow(sig01) #383

sig$Gene<-rownames(sig)
write.table(sig, "20Nov25_DE_results_postcop.txt", quote=F, row.names=F, col.names=T, sep="\t")
```


#Gene Set Enrichment Analysis
################
# GSEA
################

```{r}
#BiocManager::install("enrichplot")
#BiocManager::install("clusterProfiler")
```


```{r}
library(clusterProfiler)
library(enrichplot)
library(biomaRt)
```

```{r}
# Connect to Ensembl BioMart and select your organism (e.g., Papio anubis)
ensembl <- useEnsembl("ensembl", dataset = "panubis_gene_ensembl", mirror = "useast") 

# Retrieve GO term annotations for your genes
go_annotations <- getBM(attributes = c("external_gene_name", "go_id", "name_1006"),
                        mart = ensembl)

# Create TERM2GENE data frame (GO term to gene symbol mapping)
TERM2GENE <- go_annotations[, c("go_id", "external_gene_name")]

# Create TERM2NAME data frame (GO term to GO name mapping)
TERM2NAME <- unique(go_annotations[, c("go_id", "name_1006")])
```


```{r}
coeff_play<-as.data.frame(fit2$coeff)
coeff_play$geneID<-rownames(coeff_play)
```

```{r}
original_PostcopY_gene_list <- (coeff_play$PostcopY)
names(original_PostcopY_gene_list) <- coeff_play$geneID
PostcopY_gene_list<-na.omit(original_PostcopY_gene_list)
PostcopY_gene_list = sort( (PostcopY_gene_list), decreasing = TRUE)
```

```{r}
gsea_PostcopY <- GSEA(
  geneList = PostcopY_gene_list,
  TERM2GENE = TERM2GENE,
  TERM2NAME = TERM2NAME,
  pvalueCutoff = 0.05
)
```

#save results
```{r}
foldchange_df <- data.frame(
  gene = names(PostcopY_gene_list),
  fold_change = PostcopY_gene_list
)

write.table(foldchange_df, "PostcopY_cnetplot_foldchanges.txt", quote=F, sep="\t")
write.table(gsea_PostcopY@result, "gsea_Postcop_robust.txt", quote=F, row.names=F, col.names=T, sep="\t")
```

```{r}
Postcop_cnet<-cnetplot(gsea_PostcopY, layout= "kk", showCategory=c("inflammatory response", "cellular response to lipopolysaccharide"), foldChange=PostcopY_gene_list, color_category = "gray40") +
  scale_color_gradient(low="gray80", high="#76D6FF");Postcop_cnet

ggsave("18Nov25_Postcop_cnet_plot.pdf", Postcop_cnet, width=7, height=4)
```

## Make a heatmap of genes that are part of those GO terms
```{r}
immune_genes<-c("IL1RN","MAP2K3","BCL6","NFKBIA","NCF1","PIK3CD","ADAM8","CMKLR1","MEFV","NAMPT","TLR2","TNFAIP3", "CXCL8","AIM2","CD14","S100A9","TNIP1","NFKBIZ","TCIRG1","IL1B","TLR1", "SLC11A1","FOXP3","TLR6","NAGLU", "ZFP36", "RARA", "MAPK14", "NOD2", "JAK2", "IRAK3") 

sig_immune_genes<-subset(sig, rownames(sig) %in% immune_genes) %>%
  unique()

norm_counts<-voom_matrix$E
fitted_counts<-fitted(fit2)
colnames(fitted_counts)<- colnames(norm_counts)
immune_gene_resid<-as.data.frame(fitted_counts[rownames(fitted_counts) %in% rownames(sig_immune_genes),])
immune_gene_resid_reversed <- immune_gene_resid[, rev(colnames(immune_gene_resid))]

Sig_immune_genes_heatmap<-pheatmap(immune_gene_resid_reversed,
         color = colorRampPalette(c("#531B93","white", "#76D6FF"))(100),
         border_color = "white",
         cluster_cols = FALSE, 
         cluster_rows = FALSE,
         scale="row",
         show_colnames=FALSE); Sig_immune_genes_heatmap

ggsave("18Nov25_Sig_immune_genes_heatmap.pdf", Sig_immune_genes_heatmap, width=6, height=3.5)

write.table(fitted_counts, "postcop_fittedcounts.txt", quote=F, sep="\t")
```


# plot gene that is upreg in postcop samples
```{r}
#Figure 2C
#TLR2
norm_counts<-voom_matrix$E
TLR2_counts<-as.data.frame(norm_counts["TLR2",])
colnames(TLR2_counts)<-c("norm_counts")
TLR2_counts$sampleID<-rownames(TLR2_counts)
rownames(TLR2_counts)<-c(1:55)
TLR2_counts$Postcop<- "Yes"
TLR2_counts[c(26:55),"Postcop"]<-"No"
TLR2_counts$gene<- "TLR2"

TLR2_plot<-ggplot(data=TLR2_counts, aes(x=Postcop, y=norm_counts)) +
  geom_boxplot(aes(color=Postcop), outlier.shape=NA) +
  geom_jitter(width=0.2, size=0.5, alpha=0.5)+
    scale_color_manual(values= c("#531B93", "#D783FF"), labels= c("No", "Yes")) +
  theme_light() +
  theme(legend.position="none", plot.title = element_text(hjust = 0.5)) +
  ylab("normalized counts") +
  xlab("Post-copulatory status") +
  ggtitle("TLR2"); TLR2_plot

ggsave("Postcop_TLR2.pdf", TLR2_plot, width=2.3, height=3)

```
